### ISSUE TITLE
Rule Proposal: no-strict-with-complex-params

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#6377]
```js
function foo(a = 1) {
    "use strict"
}
```

This is a valid code in ES2015, but this is an invalid code (syntax error) in ES2016.

> It is a Syntax Error if ContainsUseStrict of FunctionBody is true and IsSimpleParameterList of FormalParameters is false.
> https://tc39.github.io/ecma262/2016/#sec-function-definitions-static-semantics-early-errors

So a combination of a non-simple parameter list and a `"use strict"` directive is going to break the software in future.

The purpose of this rule is that it protects software from the break.

```js
/*eslint no-strict-with-complex-params: error*/

// ✘ BAD
function foo(a = 1) {
    "use strict"
}
function foo({a, b}) {
    "use strict"
}

// ✔ GOOD
function foo(a, b, c) {
    "use strict"
}
function foo(a = 1) {
}
function foo({a, b}) {
}
```

I think this is a good rule for `eslint:recommended`, so I separated this rule proposal from `strict` rule.
I'm happy to work on this if accepted.

**Comment by platinumazure** [comment]
Question: Should these examples warn too?

```js
(function () {
    "use strict";

    function myFunc(a=1) {
        // No explicit use strict here but we are already in strict mode
    }
}());

(function () {
    "use strict";

    function myFunc({a, b}) {
        // No explicit use strict here but we are already in strict mode
    }
}());

(function () {
    "use strict";

    function myFunc(a) {
        // Simple parameter list. No explicit use strict here but we are already in strict mode
    }
}());
```

Thanks!

**Comment by mysticatea** [comment]
No, there is no function which has both a `"use strict"` directive and a non-simple parameter in that examples.

**Comment by platinumazure** [comment]
So it only applies to the actual function body and not to functions that
are strict by containing scope. Good to know, thanks!

**Comment by mysticatea** [comment]
Yes, this is not related in outer scopes.

```js
(function () {
    "use strict";

    function myFunc(a=1) { // this is syntax error.
        "use strict"
    }

    function myFunc(a=1) { // this is OK.
    }
}());

(function () {
    function myFunc(a=1) { // this is syntax error.
        "use strict"
    }

    function myFunc(a=1) { // this is OK.
    }
}());
```

And arrow functions, generator functions, object method shorthand, and class methods have the same behavior.

Actually, Chrome 51 does this behavior, we can try it in devtools' console.

**Comment by michaelficarra** [comment]
This doesn't need to be a rule. It's a syntax error, so if the parser accepts it, there's a bug in the parser.

**Comment by platinumazure** [comment]
@michaelficarra That's only true if the parser is set to parse ES2016, no?
ES2015 allows for that structure. The purpose of the rule (from what I can
tell) is to help users with the transition before they switch.

That said, maybe this would be better served as a plugin rule.

**Comment by michaelficarra** [comment]
Do we have precedent for any other rules like that? If I recall correctly, espree disallows valid ES5 constructs like for-in with initialisers and identifiers named `await` (even though that's only supposed to be reserved in modules), and doesn't provide linting rules to catch them. I'd love to be wrong here, though.

**Comment by ilyavolodin** [comment]
As far as I understand, if you have parser setup as ES5, there's no way to create a situation where this would happen. @mysticatea correct me if I'm wrong, but you need one of the ES6 constructs as arguments of a function to have this problem? If that's the case, and if Espree will catch it, I'm with @michaelficarra and we don't need a rule for this, as it will be a parser error.

**Comment by platinumazure** [comment]
@ilyavolodin Just to clarify, while @michaelficarra brought up ES5 constructs as an example of potential analogies, I'm talking about ES6 vs ES7.

Someone with ES6
* Can use the ES6 arguments constructs, and
* Can use "use strict" in function bodies,
* Without causing a parsing error if ecmaVersion=6. (If I'm understanding correctly.)

**Comment by michaelficarra** [comment]
Ah, that's right @ilyavolodin. ES5 only allowed for simple parameter lists, and this syntax error only occurs for non-simple parameter lists.

**Comment by ilyavolodin** [comment]
@platinumazure It should be a syntax error. It is in Chrome, for example. Based on the text @mysticatea pasted, it also should be the correct behavior per spec. So we should make this change to Espree, not a separate rule for ESLint.
@michaelficarra Why did you cross out your answer? Is that statement incorrect? Is there a construct with ES5 syntax and `use strict` in the body of the function that should not be allowed?

**Comment by michaelficarra** [comment]
Because we're talking about distinguishing ES2015 and ES2016. My statement was correct but irrelevant right now.

**Comment by mysticatea** [comment]
Actually, I reported this to also Acorn: https://github.com/ternjs/acorn/issues/423
Espree is using Acorn, so we will get a syntax error on this if it's in `ecmaVersion: 7` setting.
But if we want to get a syntax error in also `ecmaVersion: 6` setting, we would need to override Acorn 's behavior.

And, yeah, I feel that a rule to catch matters which don't work in modules for scripts is a good idea!

**Comment by mysticatea** [comment]
@ilyavolodin Chrome has supported ES2016 already. We can use `**` operator as well.

**Comment by ilyavolodin** [comment]
Ah, I see. 2015 -> ES6 was throwing me off. So this is only a syntax error in ecmaVersion 7, but not in ecmaVersion 6, right?

**Comment by mysticatea** [comment]
Yes!

**Comment by mysticatea** [comment]
May I be given :+1:s here?

**Comment by ilyavolodin** [comment]
@mysticatea Just to summarize.

This is a syntax error with ES2016, however it's valid code in ES2015. You want to create this rule to flag those constructs in ES2015 to prevent issues when browsers/node will upgrade their engine to support ES2017. Is that correct?

I'm a bit on the fence with this. It seems like a valid issue, but I'm not sure it's big/important enough to include it into the core. @eslint/eslint-team thoughts?

**Comment by mikesherov** [comment]
I don't think this should be a rule. `ecmaVersion` controls valid syntax and I don't think we need to call this out as a separate rule.

:-1:

**Comment by mysticatea** [comment]
@ilyavolodin I's correct. Then I think it's important because two reasons.

- To image that code is invalid is difficult for many people, I guess. e.g. in https://github.com/ternjs/acorn/issues/423, he wrote it's innocent-looking form.
- ESLint has been misleading people to wrong if they are using [`strict: function`](http://eslint.org/docs/rules/strict#function) configuration.

Granted, this problem is temporal until people come to not use ES6.
But we still can configure ESLint with `ecmaVersion: 6`.

@mikesherov This is to warn a deprecated syntax to protect us from the break. That code is valid in ES6, but invalid in ES7 or later, so it would break software in future. How does `ecmaVersion` control it? Should `ecmaVersion: 6` raises a syntax error on the valid code?

**Comment by nzakas** [comment]
There's actually no way this rule would run, assuming Espree throws a syntax error. I'd say let's check that Espree does actually throw an error in this case. If not, let's file an issue there.

**Comment by mysticatea** [comment]
@nzakas There is an issue: https://github.com/ternjs/acorn/issues/423
That code is a valid code in ES6,  no syntax error. But it's invalid in ES7 or later.

**Comment by nzakas** [comment]
Ah, I see. I got it now. 

Here's the problem I see : we don't currently have other rules that are "future syntax problems." We could start doing that, of course, but who is the target audience? Devs who are using ES6 now will probably move along to ES7 and then ES8 pretty rapidly, so this rule wouldn't help them. Those stuck in ES5 can't use default parameters, so it doesn't help them either. So we would be creating a rule that starts with a small audience and shrinks over time. 

Is that something worth doing? I have a hard time saying yes.

**Comment by mysticatea** [comment]
I cannot show the number of people which are using `env: {es6: true}` or `parserOptions: {ecmaVersion: 6}`. We are not collecting statistics data....

**Comment by nzakas** [comment]
I know, I'm not asking for stats. I'm asking who are the group of people you think need this rule?

**Comment by mysticatea** [comment]
All people which using `es6` setting. 
Especially, people which are using `strict` rule with `function` option.

**Comment by ilyavolodin** [comment]
@nzakas I can see how somebody would be writing an es6 code right now, and then once Node 8 is released (or whatever version will start supporting es7) they will start getting syntax errors all over their code, because our rule told them to put `use strict` inside functions.
But I'm still not sure about having this inside the core. It seems like the rule will be useful for a short period of time, until es7 is supported everywhere. And then it will be a syntax error, and I don't think we need a rule that will not ever work (since espree will fail to parse the code).
Maybe we should create an option for `strict` rule that would not require `use strict` inside function that have es6 complex constructs as arguments? Or maybe we should enhance `safe` option to not do that by default?

**Comment by mikesherov** [comment]
Ah, we should fix the `use strict` rule to not suggest when default parameters are in use. Perhaps that's the correct place to account for this.

**Comment by nzakas** [comment]
I like the idea of modifying `strict` to account for this.

**Comment by nzakas** [comment]
FWIW, this check is already in Chrome, so it throws an error if you use `"use strict"` in a function definition with non-simple parameter list.

**Comment by mysticatea** [comment]
As mentioned in my first post, the reason that I separated this proposal from `strict` rule is that I thought this is good one for `eslint:recommended`. This reports clearly errors.

> FWIW, this check is already in Chrome, so it throws an error if you use "use strict" in a function definition with non-simple parameter list.

Yes, Chrome has supported es7 already.
But I guess many people is using ESLint with es6 configuration.

**Comment by mysticatea** [comment]
OK, this is a rare case, so it's not proper for a new rule.

If `strict` rule addresses this, may I modify the default behavior?
Or does it need a new option?

**Comment by mikesherov** [comment]
To me this absolutely must modify default behavior, but because it's essentially a bug fix that might start failing builds, it'd need to go out in a minor release.

**Comment by pedrottimark** [comment]
Asking to confirm things not discussed explicitly:

* The reason to configure `strict` rule with `function` option is because the code is **neither** in ECMAScript module **nor** CommonJS module?
* In a function which has a syntax restriction against strict mode directive, the code in its body **cannot** run with strict mode semantics **unless** it is implied from the outer scope, either explicitly or implicitly?
* The design of the language is leaning away from strict mode directive because modules are coming?
* So, the preferred solution for code running in a script (for example, non-bundled code in browser environment) is wrap code in immediately-invoked function expression with strict mode directive?

**Comment by mikesherov** [comment]
> The reason to configure strict rule with function option is because the code is neither in ECMAScript module nor CommonJS module?

Correct. Although the caveat is that CommonJS modules in node  are *already* parsed as function body. It's implicitly in an IIFE

> In a function which has a syntax restriction against strict mode directive, the code in its body cannot run with strict mode semantics unless it is implied from the outer scope, either explicitly or implicitly?

Correct.

> The design of the language is leaning away from strict mode directive because modules are coming?

Modules are here :-). Loaders aren't ;-). But that's correct. Modules are always strict anyway. 

> So, the preferred solution for code running in a script (for example, non-bundled code in browser environment) is wrap code in immediately-invoked function expression with strict mode directive?

Correct. This has always been the advice: hoist "use strict" as high as possible without hoisting it to global scope (because 3rd party code may be sloppy). This implies an outermost IIFE, which is a best practice anyway, to avoid creating globals in your code.

**Comment by alberto** [comment]
Is this a bug fix or a new feature? If it is a feature, changing its default would be a breaking change.

**Comment by pedrottimark** [comment]
Thank you for the answers in https://github.com/eslint/eslint/issues/6377#issuecomment-226181444

Given the software quality goal of the `strict` rule, then would allowing the absence of a strict mode directive for the `function` option in this situation be like replacing a blown fuse with a copper penny because that function body would not have strict mode semantics given the config?

What do you think about the following solution:
* The rule continues to report a problem, therefor no breaking change.
* The rule provides a more specific message about absence of strict mode directive for `function` option in functions which have the syntax restriction if `"ecmaVersion": 6`
* The rule doc explains this situation so you can act on the rule error in ES6/2015 or the parser error in ES7/2016. We can build on recent precedent in #5991 to describe problems that might be either rule errors or parser errors depending on the configuration.

**Comment by mysticatea** [comment]
@alberto there are 2 matters here:

1. `strict` rule with `function` option is warning `function foo(a = 1) {}`, this would lead people to a syntax error, so I think this is a bug (refs #6405) and it's a `"semver: patch"` matter.
2. `strict` rule should warn `function foo(a = 1) { "use strict" }` since it's going to be broken. This is an enhancement. My question is which is better that this is the default behavior or this is hidden behind an option. I think changing the default behavior is proper since the warning is critical.

@pedrottimark Sounds good. Indeed, if `strict` rule with `function` option does not warn `function foo(a = 1) {}`, people can't get warnings for non-strict mode functions. We can keep the warning and add a more specific message.

**Comment by mikesherov** [comment]
@mysticatea, I think you got it right. My only point about part 2 is that it would be `semver: minor`, not `semver: major` because even though it's a change to default behavior, it's a bugfix that breaks builds, which we had decided in a minor.

**Comment by nzakas** [comment]
Agree with @mikesherov 

I think we should close this issue now and move the conversation over to https://github.com/eslint/eslint/issues/6405

I've also submitted a PR to Acorn to catch the syntax error. https://github.com/ternjs/acorn/pull/424

**Comment by mysticatea** [comment]
OK.
Thank you everyone for amount of discussion!

**Comment by Mouvedia** [comment]
Shouldn't this be added as a warning to `eslint:recommended`?

**Comment by mysticatea** [comment]
This was implemented in [strict](http://eslint.org/docs/rules/strict) rule. Then the `strict` rule is one of stylistic rules, so `eslint:recommended` shouldn't include it. :cry:

**Comment by platinumazure** [comment]
I'm not sure `strict` is stylistic anymore. Should we consider changing its category?

**Comment by michaelficarra** [comment]
Yes.

**Comment by ljharb** [comment]
I don't think it was ever stylistic, but with Modules and class bodies being automatically strict, not using strict mode is very unlikely to be intentional (and an inline comment is appropriate when it is).

**Comment by mysticatea** [comment]
I'm sorry, let me correct me. 
 `strict` rule isn't in stylistic category. But `eslint:recommended` does not include such rules which people should choose their preference. In this case, global, function, or never.

Though, I don't oppose if we add `strict` rule to `eslint:recommended`.
